File: server/backend/app/protocol/packet.py
from dataclasses import dataclass, field
from enum import IntEnum

from .crc16 import crc16_ccitt

VER = 0x01
GATEWAY_ID = 0x00
BROADCAST = 0xFF
FRAG_SINGLE = 0x80  # bit7=LAST, 인덱스 0
MAX_PAYLOAD = 200
_HEADER_LEN = 7
_CRC_LEN = 2


class MsgType(IntEnum):
    PING = 0x01
    PONG = 0x02
    SET_TEMPLATE = 0x10
    SET_FIELD = 0x11
    SET_QR = 0x12
    COMMIT = 0x13
    IMG_FRAG = 0x14
    ACK = 0x20
    STATUS_REQ = 0x30
    STATUS_RES = 0x31


class AckResult(IntEnum):
    OK = 0
    CRC_FAIL = 1
    BUSY = 2
    BAD_TYPE = 3


class PacketError(Exception):
    pass


class CrcError(PacketError):
    pass


@dataclass(frozen=True)
class Packet:
    src: int
    dst: int
    type: MsgType
    seq: int
    payload: bytes = field(default=b"")
    frag: int = FRAG_SINGLE
    ver: int = VER


def encode(p: Packet) -> bytes:
    if len(p.payload) > MAX_PAYLOAD:
        raise PacketError(f"payload {len(p.payload)}B > {MAX_PAYLOAD}B")
    body = bytes([p.ver, p.src, p.dst, p.type, p.seq, p.frag, len(p.payload)]) + p.payload
    return body + crc16_ccitt(body).to_bytes(_CRC_LEN, "little")


def decode(buf: bytes) -> Packet:
    if len(buf) < _HEADER_LEN + _CRC_LEN:
        raise PacketError("buffer too short")
    length = buf[6]
    if len(buf) != _HEADER_LEN + length + _CRC_LEN:
        raise PacketError("LEN mismatch")
    body, crc = buf[:-_CRC_LEN], int.from_bytes(buf[-_CRC_LEN:], "little")
    if crc16_ccitt(body) != crc:
        raise CrcError("CRC16 mismatch")
    try:
        msg_type = MsgType(buf[3])
    except ValueError as exc:
        raise PacketError(f"unknown TYPE 0x{buf[3]:02X}") from exc
    return Packet(src=buf[1], dst=buf[2], type=msg_type, seq=buf[4],
                  payload=bytes(buf[7:-_CRC_LEN]), frag=buf[5], ver=buf[0])


# ---- 페이로드 빌더/파서 (PROTOCOL.md §3) ----

def build_set_field(field_id: int, text: str) -> bytes:
    raw = text.encode("utf-8")
    if len(raw) > MAX_PAYLOAD - 2:
        raise PacketError("field text too long")
    return bytes([field_id, len(raw)]) + raw


def build_set_qr(url: str, qr_slot: int = 0) -> bytes:
    raw = url.encode("utf-8")
    if len(raw) > MAX_PAYLOAD - 2:
        raise PacketError("qr url too long")
    return bytes([qr_slot, len(raw)]) + raw


def build_ack(ack_seq: int, result: AckResult) -> bytes:
    return bytes([ack_seq, result])


def parse_ack(payload: bytes) -> tuple[int, AckResult]:
    return payload[0], AckResult(payload[1])


def build_pong(batt_mv: int, rssi: int, status: int) -> bytes:
    return batt_mv.to_bytes(2, "little") + rssi.to_bytes(1, "little", signed=True) \
        + bytes([status])


def parse_pong(payload: bytes) -> tuple[int, int, int]:
    return (int.from_bytes(payload[0:2], "little"),
            int.from_bytes(payload[2:3], "little", signed=True), payload[3])


def build_status_res(batt_mv: int, last_seq: int, uptime_s: int, err_cnt: int) -> bytes:
    return (batt_mv.to_bytes(2, "little") + bytes([last_seq])
            + uptime_s.to_bytes(2, "little") + bytes([err_cnt]))


def parse_status_res(payload: bytes) -> tuple[int, int, int, int]:
    return (int.from_bytes(payload[0:2], "little"), payload[2],
            int.from_bytes(payload[3:5], "little"), payload[5])

File: shared/lib/efb_protocol/src/packet.cpp
#include <efb/packet.h>

#include <string.h>

#include <efb/crc16.h>

namespace efb {

namespace {

void put_u16_le(uint8_t* p, uint16_t v) {
    p[0] = static_cast<uint8_t>(v & 0xFF);
    p[1] = static_cast<uint8_t>(v >> 8);
}

uint16_t get_u16_le(const uint8_t* p) {
    return static_cast<uint16_t>(p[0] | (static_cast<uint16_t>(p[1]) << 8));
}

}  // namespace

bool is_known_type(uint8_t type) {
    switch (type) {
        case PING:
        case PONG:
        case SET_TEMPLATE:
        case SET_FIELD:
        case SET_QR:
        case COMMIT:
        case IMG_FRAG:
        case RESET:
        case ACK:
        case STATUS_REQ:
        case STATUS_RES:
            return true;
        default:
            return false;
    }
}

size_t encode(const Packet& p, uint8_t* out, size_t out_cap) {
    if (p.len > MAX_PAYLOAD) return 0;
    const size_t total = HEADER_LEN + p.len + CRC_LEN;
    if (out_cap < total) return 0;

    out[0] = p.ver;
    out[1] = p.src;
    out[2] = p.dst;
    out[3] = p.type;
    out[4] = p.seq;
    out[5] = p.frag;
    out[6] = p.len;
    if (p.len) memcpy(out + HEADER_LEN, p.payload, p.len);

    const uint16_t crc = crc16_ccitt(out, HEADER_LEN + p.len);  // CRC 대상 = VER..PAYLOAD
    put_u16_le(out + HEADER_LEN + p.len, crc);
    return total;
}

PacketErr decode(const uint8_t* buf, size_t len, Packet& out) {
    if (len < HEADER_LEN + CRC_LEN) return PacketErr::TOO_SHORT;

    const uint8_t payload_len = buf[6];
    if (len != HEADER_LEN + static_cast<size_t>(payload_len) + CRC_LEN) {
        return PacketErr::LEN_MISMATCH;
    }

    const size_t body_len = HEADER_LEN + payload_len;
    if (crc16_ccitt(buf, body_len) != get_u16_le(buf + body_len)) return PacketErr::CRC;
    if (!is_known_type(buf[3])) return PacketErr::UNKNOWN_TYPE;

    out.ver = buf[0];
    out.src = buf[1];
    out.dst = buf[2];
    out.type = buf[3];
    out.seq = buf[4];
    out.frag = buf[5];
    out.len = payload_len;
    if (payload_len) memcpy(out.payload, buf + HEADER_LEN, payload_len);
    return PacketErr::NONE;
}

size_t build_set_field(uint8_t field_id, const char* text, size_t text_len, uint8_t* out,
                       size_t out_cap) {
    if (text_len > MAX_TEXT || out_cap < text_len + 2) return 0;
    out[0] = field_id;
    out[1] = static_cast<uint8_t>(text_len);
    if (text_len) memcpy(out + 2, text, text_len);
    return text_len + 2;
}

size_t build_set_qr(uint8_t qr_slot, const char* url, size_t url_len, uint8_t* out,
                    size_t out_cap) {
    if (url_len > MAX_TEXT || out_cap < url_len + 2) return 0;
    out[0] = qr_slot;
    out[1] = static_cast<uint8_t>(url_len);
    if (url_len) memcpy(out + 2, url, url_len);
    return url_len + 2;
}

size_t build_ack(uint8_t ack_seq, uint8_t result, uint8_t* out, size_t out_cap) {
    if (out_cap < 2) return 0;
    out[0] = ack_seq;
    out[1] = result;
    return 2;
}

size_t build_pong(uint16_t batt_mv, int8_t rssi, uint8_t status, uint8_t* out, size_t out_cap) {
    if (out_cap < 4) return 0;
    put_u16_le(out, batt_mv);
    out[2] = static_cast<uint8_t>(rssi);
    out[3] = status;
    return 4;
}

size_t build_status_res(uint16_t batt_mv, uint8_t last_seq, uint16_t uptime_s, uint8_t err_cnt,
                        uint8_t* out, size_t out_cap) {
    if (out_cap < 6) return 0;
    put_u16_le(out, batt_mv);
    out[2] = last_seq;
    put_u16_le(out + 3, uptime_s);
    out[5] = err_cnt;
    return 6;
}

bool parse_ack(const uint8_t* payload, size_t len, uint8_t& ack_seq, uint8_t& result) {
    if (len < 2) return false;
    ack_seq = payload[0];
    result = payload[1];
    return true;
}

bool parse_pong(const uint8_t* payload, size_t len, uint16_t& batt_mv, int8_t& rssi,
                uint8_t& status) {
    if (len < 4) return false;
    batt_mv = get_u16_le(payload);
    rssi = static_cast<int8_t>(payload[2]);
    status = payload[3];
    return true;
}

bool parse_status_res(const uint8_t* payload, size_t len, uint16_t& batt_mv, uint8_t& last_seq,
                      uint16_t& uptime_s, uint8_t& err_cnt) {
    if (len < 6) return false;
    batt_mv = get_u16_le(payload);
    last_seq = payload[2];
    uptime_s = get_u16_le(payload + 3);
    err_cnt = payload[5];
    return true;
}

}  // namespace efb
